File: modules/agent/src/services/note_service.py
"""Service layer for note business logic."""

import logging
from typing import Optional, Dict, Any, List

from fastapi import HTTPException

from repositories.note_repository import (
    find_notes_by_entity,
    find_note_by_id,
    create_note as create_note_repo,
    update_note as update_note_repo,
    delete_note as delete_note_repo,
)

logger = logging.getLogger(__name__)

ENTITY_TYPE_MAP = {
    "organization": "Organization",
    "individual": "Individual",
    "contact": "Contact",
    "lead": "Lead",
}


def _normalize_entity_type(entity_type: str) -> str:
    """Normalize entity_type to PascalCase."""
    return ENTITY_TYPE_MAP.get(entity_type.lower(), entity_type)


async def get_notes_by_entity(tenant_id: int, entity_type: str, entity_id: int) -> List:
    """Get all notes for a specific entity."""
    normalized_type = _normalize_entity_type(entity_type)
    return await find_notes_by_entity(tenant_id, normalized_type, entity_id)


async def get_note(note_id: int, tenant_id: int):
    """Get a single note by ID."""
    note = await find_note_by_id(note_id, tenant_id)
    if not note:
        raise HTTPException(status_code=404, detail="Note not found")
    return note


async def create_note(
    tenant_id: int,
    entity_type: str,
    entity_id: int,
    content: str,
    user_id: Optional[int],
):
    """Create a new note."""
    note_data = {
        "tenant_id": tenant_id,
        "entity_type": _normalize_entity_type(entity_type),
        "entity_id": entity_id,
        "content": content,
        "created_by": user_id,
    }
    return await create_note_repo(note_data)


async def update_note(
    note_id: int,
    tenant_id: int,
    content: str,
    user_id: Optional[int],
):
    """Update a note."""
    note = await update_note_repo(note_id, tenant_id, {"content": content, "updated_by": user_id})
    if not note:
        raise HTTPException(status_code=404, detail="Note not found")
    return note


async def delete_note(note_id: int, tenant_id: int):
    """Delete a note."""
    deleted = await delete_note_repo(note_id, tenant_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Note not found")
    return True

File: modules/agent-go/internal/services/note_service.go
package services

import (
	"errors"

	"github.com/pina-colada-co/agent-go/internal/models"
	"github.com/pina-colada-co/agent-go/internal/repositories"
)

var ErrNoteNotFound = errors.New("note not found")

type NoteService struct {
	noteRepo *repositories.NoteRepository
}

func NewNoteService(noteRepo *repositories.NoteRepository) *NoteService {
	return &NoteService{noteRepo: noteRepo}
}

func (s *NoteService) GetNotesByEntity(entityType string, entityID int64, tenantID *int64) ([]models.Note, error) {
	return s.noteRepo.FindByEntity(entityType, entityID, tenantID)
}

func (s *NoteService) GetNote(id int64) (*models.Note, error) {
	note, err := s.noteRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if note == nil {
		return nil, ErrNoteNotFound
	}
	return note, nil
}

type NoteCreateInput struct {
	TenantID   int64
	EntityType string
	EntityID   int64
	Content    string
	UserID     int64
}

func (s *NoteService) CreateNote(input NoteCreateInput) (*models.Note, error) {
	repoInput := repositories.NoteCreateInput{
		TenantID:   input.TenantID,
		EntityType: input.EntityType,
		EntityID:   input.EntityID,
		Content:    input.Content,
		UserID:     input.UserID,
	}
	id, err := s.noteRepo.Create(repoInput)
	if err != nil {
		return nil, err
	}
	return s.noteRepo.FindByID(id)
}

func (s *NoteService) UpdateNote(id int64, content string, userID int64) (*models.Note, error) {
	note, err := s.noteRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if note == nil {
		return nil, ErrNoteNotFound
	}
	updates := map[string]interface{}{
		"content":    content,
		"updated_by": userID,
	}
	if err := s.noteRepo.Update(id, updates); err != nil {
		return nil, err
	}
	return s.noteRepo.FindByID(id)
}

func (s *NoteService) DeleteNote(id int64) error {
	note, err := s.noteRepo.FindByID(id)
	if err != nil {
		return err
	}
	if note == nil {
		return ErrNoteNotFound
	}
	return s.noteRepo.Delete(id)
}
